gb-roms/src/controllers/mbc3.rs

```diff
@@ -0,0 +1,243 @@
+        }
+        let address = addr.get_address();
+        let ram_bank = self.regs.ram_bank;
+        match ram_bank {
+            0x0..=0x3 => Ok(self.ram_banks[ram_bank as usize][(address as usize) & 0x1FFF]),
+            0x8 => Ok(self.regs.rtc.seconds),
+            0x9 => Ok(self.regs.rtc.minutes),
+            0xA => Ok(self.regs.rtc.hours),
+            0xB => Ok(self.regs.rtc.lower_day_counter),
+            0xC => Ok(self.regs.rtc.upper_day_counter),
+            _ => Err(Error::new_segfault(addr)),
+        }
+    }
+
+    fn write_ram(&mut self, v: u8, addr: Box<dyn Address<Area>>) -> Result<(), Error> {
+        if !self.regs.ram_enabled {
+            return Err(Error::new_segfault(addr));
+        }
+        let address = addr.get_address();
+        let ram_bank = self.regs.ram_bank;
+        match ram_bank {
+            0x0..=0x3 => self.ram_banks[ram_bank as usize][(address as usize) & 0x1FFF] = v,
+            0x8 => self.regs.rtc.seconds = v,
+            0x9 => self.regs.rtc.minutes = v,
+            0xA => self.regs.rtc.hours = v,
+            0xB => self.regs.rtc.lower_day_counter = v,
+            0xC => self.regs.rtc.upper_day_counter = v,
+            _ => return Err(Error::new_segfault(addr)),
+        }
+        Ok(())
+    }
+}
+
+impl FileOperation<Area> for MBC3 {
+    fn read(&self, addr: Box<dyn Address<Area>>) -> Result<u8, Error> {
+        match addr.area_type() {
+            Area::Rom => self.read_rom(addr),
+            Area::Ram => self.read_ram(addr),
+            _ => Err(Error::new_bus_error(addr)),
+        }
+    }
+    fn write(&mut self, v: u8, addr: Box<dyn Address<Area>>) -> Result<(), Error> {
+        match addr.area_type() {
+            Area::Rom => self.write_rom(v, addr),
+            Area::Ram => self.write_ram(v, addr),
+            _ => Err(Error::new_bus_error(addr)),
+        }
+    }
+}
+
+#[derive(Default)]
+struct MBC3Regs {
+    rom_bank: u8,
+    ram_bank: u8,
+    ram_enabled: bool,
+    rtc: RTCRegs,
+    last_writed_byte: Option<u8>,
+}
+
+#[derive(Default)]
+struct RTCRegs {
+    seconds: u8,
+    minutes: u8,
+    hours: u8,
+    lower_day_counter: u8,
+    upper_day_counter: u8,
+}
+
+impl<T: ReadRtcRegisters> From<&T> for RTCRegs {
+    fn from(clock: &T) -> Self {
+        Self {
+            seconds: clock.seconds(),
+            minutes: clock.minutes(),
+            hours: clock.hours(),
+            lower_day_counter: clock.lower_days(),
+            upper_day_counter: clock.control(),
+        }
+    }
+}
+
+impl From<Vec<u8>> for RTCRegs {
+    fn from(data: Vec<u8>) -> Self {
+        if data.len() < 5 {
+            return RTCRegs::default();
+        }
+        Self {
+            seconds: data[0],
+            minutes: data[1],
+            hours: data[2],
+            lower_day_counter: data[3],
+            upper_day_counter: data[4],
+        }
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct Mbc3Data {
+    rtc: Vec<u8>,
```

Comment:
use directly the clock (here `Naive`) not a vector

Reply:
Naive doesn't implement `serde::Serialize, serde::Deserialize`, it seems that I can't use this type in struct `Mbc3Data`

Reply:
you can implement the trait for `Naive`

Reply:
But then Instant doesn't implement it